src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -3480,6 +3468,23 @@ public override async Task<RachisConnection> ConnectToPeer(string url, string ta
             }
         }
 
+        private async Task<TcpConnectionHeaderMessage.SupportedFeatures> NegotiateProtocolVersionAsyncForCluster(string url, TcpConnectionInfo info, Stream stream, JsonOperationContext ctx, string tag)
+        {
+            var parameters = new AsyncTcpNegotiateParameters
+            {
+                Database = null,
+                Operation = TcpConnectionHeaderMessage.OperationTypes.Cluster,
+                Version = TcpConnectionHeaderMessage.ClusterTcpVersion,
+                ReadResponseAndGetVersionCallbackAsync = ClusterReadResponseAndGetVersion,
+                DestinationUrl = url,
+                DestinationNodeTag = tag,
+                SourceNodeTag = _parent.Tag,
+                DestinationServerId = info.ServerId
+            };
+
+            return await TcpNegotiation.NegotiateProtocolVersionAsync(ctx, stream, parameters);
```

Comment:
avoid the `await`, just return `Tsak`

Reply:
But then it conflicts because the return type of NegotiateProtocolVersionAsync is ValueTask and not Task. And I'm not sure if I should make the return type of this whole callback function ValueTask

Reply:
You can, but then you need to be sure that this is awaited immediately in each place and not assigned to a variable

Reply:
Not a major problem, this should be rarely called.